python/ray/data/tests/test_dataset.py

```diff
@@ -241,6 +242,43 @@ def test_dataset_lineage_serialization_in_memory(shutdown_only, lazy):
     with pytest.raises(ValueError):
         ds.serialize_lineage()
 
+    # In-memory data source unions not supported.
+    ds = ray.data.from_items(list(range(10)))
+    ds = maybe_lazy(ds, lazy)
+    ds1 = ray.data.from_items(list(range(10, 20)))
+    ds2 = ds.union(ds1)
+
+    with pytest.raises(ValueError):
+        ds2.serialize_lineage()
+
+    # Post-lazy-read unions not supported.
+    ds = ray.data.range(10).map(lambda x: x + 1)
+    ds = maybe_lazy(ds, lazy)
+    ds1 = ray.data.range(20).map(lambda x: 2 * x)
+    ds2 = ds.union(ds1)
+
+    with pytest.raises(ValueError):
+        ds2.serialize_lineage()
+
+    # Lazy read unions supported.
```

Comment:
By default they are not lazy, no?

Reply:
Yes! This is emphasizing that we're not doing any eager operations (e.g. mapping) after reading, before unioning.